### PR TITLE
CLICKHOUSE-3893 - Add modificator rollup.

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: dbms/tests/queries/0_stateless/00701_rollup.sql [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+INSERT INTO test.rollup  VALUES('a', 2, 20);
+INSERT INTO test.rollup  VALUES('a', 2, 25);
+INSERT INTO test.rollup  VALUES('b', 1, 10);
+INSERT INTO test.rollup  VALUES('b', 1, 5);
+INSERT INTO test.rollup  VALUES('b', 2, 20);
+INSERT INTO test.rollup  VALUES('b', 2, 15);
+
+SELECT a, b, sum(s), count() from test.rollup GROUP BY ROLLUP(a, b) ORDER BY a, b;
+
+SELECT a, b, sum(s), count() from test.rollup GROUP BY ROLLUP(a, b) WITH TOTALS ORDER BY a, b;
```
Comment (zhang2014) [inline dbms/tests/queries/0_stateless/00701_rollup.sql]: Why not `GROUP BY a, b WITH ROLLUP`? This benefit is friendly to people from mysql : )

Comment (alexey-milovidov) [reply dbms/tests/queries/0_stateless/00701_rollup.sql]: We will add both variants (MySQL style and standard SQL style).

**Comment by alexey-milovidov** [comment]
Submodules are updated accidentially.

File: dbms/src/Interpreters/Aggregator.h [hunk 2]
```diff
@@ -844,7 +844,7 @@ struct AggregatedDataVariants : private boost::noncopyable
                 throw Exception("Unknown aggregated data variant.", ErrorCodes::UNKNOWN_AGGREGATED_DATA_VARIANT);
         }
     }
-
+
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/Aggregator.h]: .

File: dbms/src/Parsers/ParserSelectQuery.cpp [hunk 3]
```diff
@@ -113,14 +115,35 @@ bool ParserSelectQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
...
+        if (s_rollup.ignore(pos, expected))
+            select_query->group_by_with_rollup = true;
+        else if (s_totals.ignore(pos, expected))
+            select_query->group_by_with_totals = true;
+        else
+            return false;
     }
 
-    /// WITH TOTALS
+     /// WITH TOTALS
```
Comment (alexey-milovidov) [inline dbms/src/Parsers/ParserSelectQuery.cpp]: .

File: dbms/src/Interpreters/InterpreterSelectQuery.cpp [hunk 4]
```diff
@@ -846,9 +861,10 @@ void InterpreterSelectQuery::executeAggregation(Pipeline & pipeline, const Expre
         if (descr.arguments.empty())
             for (const auto & name : descr.argument_names)
                 descr.arguments.push_back(header.getPositionByName(name));
-
+        
+                
     const Settings & settings = context.getSettingsRef();
-
+
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterSelectQuery.cpp]: .

File: dbms/src/Interpreters/InterpreterSelectQuery.cpp [hunk 5]
```diff
@@ -846,9 +861,10 @@ void InterpreterSelectQuery::executeAggregation(Pipeline & pipeline, const Expre
         if (descr.arguments.empty())
             for (const auto & name : descr.argument_names)
                 descr.arguments.push_back(header.getPositionByName(name));
-
+
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterSelectQuery.cpp]: .

File: dbms/src/Interpreters/InterpreterSelectQuery.cpp [hunk 6]
```diff
@@ -546,9 +555,12 @@ void InterpreterSelectQuery::executeImpl(Pipeline & pipeline, const BlockInputSt
             else
             {
                 need_second_distinct_pass = query.distinct && pipeline.hasMoreThanOneStream();
-
+
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterSelectQuery.cpp]: .

File: dbms/src/Interpreters/InterpreterSelectQuery.cpp [hunk 7]
```diff
@@ -955,15 +971,42 @@ void InterpreterSelectQuery::executeHaving(Pipeline & pipeline, const Expression
...
+    const Settings & settings = context.getSettingsRef();
+ 
+    Aggregator::Params params(header, keys, aggregates,
+        false, settings.max_rows_to_group_by, settings.group_by_overflow_mode,
+        settings.compile ? &context.getCompiler() : nullptr, settings.min_count_to_compile,
+        SettingUInt64(0), SettingUInt64(0),
+        settings.max_bytes_before_external_group_by, settings.empty_result_for_aggregation_by_empty_set,
+        context.getTemporaryPath());
+
+      pipeline.firstStream() = std::make_shared<RollupBlockInputStream>(pipeline.firstStream(), params);
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterSelectQuery.cpp]: Indent.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Submodules have leaked.

**Pull request merged**